App Health Extension v2: Support TLS 1.3

klugorosado opened the pull request:
### Changes
- Included Support for TLS 1.3
- Minimum TLS Version 1.0
- Max TLS Version: Default
- Added Integration Tests to ensure TLS Backward combability to avoid regressions.
- Parallelize integration tests. Testing time went down from around 25 minutes to around 2 minutes. 
- Moved Integration tests to Parallel and Sequential Folder Respectively. 
- Added Cleanup bash function on test_helper
- GitHub Workflow will Retry Flaky Integration Tests
- Added Total Time and Command Trace for GitHub Workflow to ensure all Tests are running. 
- GitHub Workflow will Run on master and feature branch
- Branching Strategy (Dev|Feature) -> master -> release

<details><summary>Details</summary>
<p>

This pull request includes a variety of changes to improve the functionality and usability of the codebase. The most important changes include adding new integration tests for testing different security protocols, adding new functions to `test_helper.bash` for cleaning up test containers and images, and modifying the `NewHttpHealthProbe` function to support TLS 1.0.

Main interface changes:

* <a href="diffhunk://#diff-2422cb6e5f570a2a3eeb5388f7e0fcc644727dfd0d34911de35e83a268f1d2efL74">`main/health.go`</a>: Added a new function to construct a URL string and modified the `NewHttpHealthProbe` function to support TLS 1.0. <a href="diffhunk://#diff-2422cb6e5f570a2a3eeb5388f7e0fcc644727dfd0d34911de35e83a268f1d2efL74">[1]</a> <a href="diffhunk://#diff-2422cb6e5f570a2a3eeb5388f7e0fcc644727dfd0d34911de35e83a268f1d2efL11-R14">[2]</a>
* <a href="diffhunk://#diff-db5ccf1b6f97761a2e7e8cb4c16754861128734a0bda7520aa4b7d51b530cbe2R216-R231">`integration-test/test/test_helper.bash`</a>: Added new functions for cleaning up test containers and images, generating unique Docker image names, and creating/deleting SSL certificates for testing purposes.

Testing improvements:

* `integration-test/test/parallel/6_tls-config.bats`: Added new integration tests for testing different security protocols to check if the application health extension is working correctly. (changes in `integration-test/test/parallel/6_tls-config.bats`)
* <a href="diffhunk://#diff-77c13d6986727b4b39067ec6227ed255794736d6bb875c3a3b294a043d7c3b83R1-R87">`main/health_test.go`</a>: Added new test cases for `NewHttpHealthProbe` and `constructAddress` functions to cover different scenarios for request path and non-default ports.

Configuration improvements:

* <a href="diffhunk://#diff-678682767f2477de3e3c546746f8568b9a1942b2c647d32331d7e774b8ff8d9fR55-R58">`.github/workflows/go.yml`</a>: Added new steps to install `parallel`, modify the `branches` field to include all feature branches, and add two new integration test steps for sequential and parallel tests with `bats` and retry failed tests. <a href="diffhunk://#diff-678682767f2477de3e3c546746f8568b9a1942b2c647d32331d7e774b8ff8d9fR55-R58">[1]</a> <a href="diffhunk://#diff-678682767f2477de3e3c546746f8568b9a1942b2c647d32331d7e774b8ff8d9fL6-R12">[2]</a> <a href="diffhunk://#diff-678682767f2477de3e3c546746f8568b9a1942b2c647d32331d7e774b8ff8d9fL88-R117">[3]</a>

Other changes:

* <a href="diffhunk://#diff-caca036aaca1d574d9aad6849f0d9d5ca148b6204592115da2fea6aa1a953e01L1-R11">`integration-test/run.sh`</a>: Refactored to use a new function `create_certificate` instead of `openssl` commands, and reordered import statements.

</p>
</details>

frank-pang-msft commented on main/health.go:
```diff
@@ -129,31 +129,37 @@ func NewHttpHealthProbe(protocol string, requestPath string, port int) *HttpHeal
-			Timeout:       timeout,
-			Transport:     transport,
+			TLSClientConfig: &tls.Config{
+				MinVersion:         tls.VersionTLS11,
```
Not against specifying version, but I think today TLS version is probably automatically selected if we don't specify min/max versions. If TLS 1.4 comes out in the future we'll have to make the change again, do you think we should leave it that way or is there a benefit to specifying?

klugorosado replied:
Honestly, it doesn't make much difference. When TLS 1.4 is released, we'll need to upgrade appHealth from GO 1.17 to the latest version anyway. The best way to future proof it is to upgrade to the latest GO version and not set a maxVersion.

frank-pang-msft commented on main/health.go:
```diff
@@ -129,31 +129,37 @@ func NewHttpHealthProbe(protocol string, requestPath string, port int) *HttpHeal
-		p.HttpClient = &http.Client{
-			CheckRedirect: noRedirect,
-			Timeout:       timeout,
+		transport = &http.Transport{
```
Does HTTP protocol use TLS, I think only HTTPS does?

klugorosado replied:
That is correct,but It wont throw error because HTTP will ignore TLS Config.
However I will still fix.

frank-pang-msft commented on integration-test/webserver/main.go:
```diff
@@ -147,16 +148,53 @@ func healthHandler(w http.ResponseWriter, r *http.Request, arguments *[]string)
+
 func main() {
 	args := flag.String("args", "", `Example usage: '2h-valid' to send StatusCode: 200, ResponseBody: { "ApplicationHealthState": "Healthy", "CustomMetrics": "<valid json>"}`)
+	tlsVersion := flag.String("tlsVersion", "tls1.3", "TLS version to use for https server. Options: 1.1, 1.2, 1.3")
```
nit not up to date.

Also, doens't really matter but maybe should we change arg name to something else since SSL is involved?

frank-pang-msft commented on integration-test/webserver/main.go:
```diff
@@ -147,16 +148,53 @@ func healthHandler(w http.ResponseWriter, r *http.Request, arguments *[]string)
 	w.Write(respBody)
 }
 
+func tlsVersionToUse(tlsVersion string) uint16 {
```
maybe securityProtocol instead of tlsVersion?

frank-pang-msft commented on integration-test/test/test_helper.bash:
```diff
@@ -16,11 +16,20 @@ in_tmp_container() {
 }
 
 rm_container() {
+    echo "DELETE: TEST_CONTAINER: $TEST_CONTAINER"
```
So this ensures that each container we spin up for each individual test gets deleted at the end?

klugorosado replied:
This was used for debugging, I will remove this.

frank-pang-msft replied:
So will each container thatis spinned up per test get deleted at the end?

klugorosado replied:
Yes, I added logic to handle that.

frank-pang-msft commented on integration-test/run.sh:
```diff
@@ -1,3 +1,5 @@
+#!/bin/bash
+source integration-test/test/test_helper.bash
+create_certificate
+sudo bats integration-test/test --jobs 10
```
For my own understanding, what does --jobs param indicate?

klugorosado replied:
The --jobs parameter indicates the max number of parallel processes to use. [Source](https://bats-core.readthedocs.io/en/stable/usage.html#parallel-execution:~:text=0.039%22%20/%3E%0A%3C/testsuite%3E%0A%3C/testsuites%3E-,Parallel%20Execution,-%C2%B6)

frank-pang-msft commented on integration-test/test/2_handler-commands.bats:
```diff
@@ -4,14 +4,16 @@ load test_helper
 
 @test "handler command: install - creates the data dir" {
-    mk_container sh -c "fake-waagent install && sleep 2"
+    container_name="${container_name}_1"
```
is there way in BATS framework to dynamically reference a testID? Adding suffix # manually can be tedious and if we ever decide to reorder these tests it can be a mess

frank-pang-msft replied:
If container name doesn't really matter, we could put this step in setup() since i think it runs each time?

frank-pang-msft commented on .github/workflows/go.yml:
```diff
@@ -50,6 +50,10 @@ jobs:
 
+    - name: Install Parallel
+      run: |
+        sudo apt install parallel
```
What is parallel? Also I recall BATS framework mentioned something about parallel execution of tests, is that not the case?

klugorosado replied:
Parallel package is a requirement for parallel execution of bats test with the --jobs flag.

frank-pang-msft reviewed (changes requested):
Quick pass

rareddaz commented on main/health.go:
```diff
@@ -129,15 +128,18 @@ func NewHttpHealthProbe(protocol string, requestPath string, port int) *HttpHeal
-			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			TLSClientConfig: &tls.Config{
+				MinVersion:         tls.VersionTLS10,
+				MaxVersion:         tls.VersionTLS13,
```
Do we need to specify min and max version here? Can we leave it to default?

klugorosado replied:
Since we Have the Integration Tests in place, I think is safe to do so.

klugorosado replied:
I attempted to leave defaults and the integration tests did not pass. This is because the Builds are on Go 1.20 which have default inversion TLS 1.2. In this case it's better to specify the min and max Versions to ensure combability.

frank-pang-msft replied:
If we omit Min and Max version shouldn’t the app use the same version that the application uses?

frank-pang-msft commented on main/health.go:
```diff
@@ -129,15 +128,17 @@ func NewHttpHealthProbe(protocol string, requestPath string, port int) *HttpHeal
-			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+				MinVersion:         tls.VersionTLS10,
```
Can we add a comment here specifying why we specify min version but not max?

I believe it's because in Go 1.18 all http clients default to TLS 1.2, but to support all TLS protocols before that we do 1.0. Is that right?

klugorosado replied:
comment added

frank-pang-msft commented on integration-test/run.sh:
```diff
@@ -1,3 +1,7 @@
+create_certificate
+# 1_basic.bats Integration Test must be run sequentially
+
+sudo bats integration-test/test/1_basic.bats && sudo bats $(find integration-test/test/ -type f -name "*.bats" ! -name "1_basic.bats") --jobs 10 
```
Just as a sanity check, I think there is a way to run the tests with verbose logs.

Have you confirmed that the tests are executing entirely and are not just passing due to returning early?

klugorosado commented:
This pull request includes a variety of changes to improve the functionality and usability of the codebase. The most important changes include adding new tests to the `integration-test` directory, adding support for tls1.0 and trimming leading slashes in the request path, and updating multiple scripts to use a custom container name instead of a hardcoded value.

Main interface changes:

* <a href="diffhunk://#diff-2422cb6e5f570a2a3eeb5388f7e0fcc644727dfd0d34911de35e83a268f1d2efL132-R144">`main/health.go`</a>: Added support for tls1.0, trimmed leading slashes in the request path, and removed an unnecessary line in the `NewHealthProbe` function. 
* <a href="diffhunk://#diff-678682767f2477de3e3c546746f8568b9a1942b2c647d32331d7e774b8ff8d9fR51-R54">`.github/workflows/go.yml`</a>: Added three new jobs for integration testing, including sequential and parallel testing with retrying failed tests.

Testing improvements:

* <a href="diffhunk://#diff-db5ccf1b6f97761a2e7e8cb4c16754861128734a0bda7520aa4b7d51b530cbe2R9-R57">`integration-test/test/test_helper.bash`</a>: Added new functions to clean up Docker containers and images after running tests, as well as create and delete self-signed certificates for testing purposes. 
* <a href="diffhunk://#diff-8aa2da999d8c499c37122550a034e558ff32fbfd4a1683a13c59c1ba7b1d636dR7-R16">`integration-test/test/4_grace-period.bats`</a>: Added functionality to the `integration-test/test/4_grace-period.bats` script, including sourcing a helper file, using a custom container name, and replacing hardcoded values with variables. 
* <a href="diffhunk://#diff-554f373b1ba296bbbf5403dcfafb077c73217421e8bdc9a349d4ba8e33c2e8cbR7-R16">`integration-test/test/2_handler-commands.bats`</a>: Updated multiple tests in `integration-test/test/2_handler-commands.bats` to use a custom container name instead of the hardcoded value `sh`. 
* <a href="diffhunk://#diff-490ae6b7f1640e574ff99e742a2bffc435826700e81cea1a73f54ed224fd43fdL47-R49">`integration-test/test/5_custom-metrics.bats`</a>: Updated multiple tests in `integration-test/test/5_custom-metrics.bats` to use a custom container name instead of hardcoded container names. 
* `integration-test/test/3_rich-states.bats`: Introduced a variable `$container_name` to ensure consistency in the container name used throughout the `3_rich-states.bats` script. 
* <a href="diffhunk://#diff-102558af691a84d712ba881d170810ddac31f5b56b409e1dec2bd2cf0431cfdbR1-R206">`integration-test/test/6_tls-config.bats`</a>: Added new test file `integration-test/test/6_tls-config.bats` to test SSL/TLS protocols supported by the webserver.

Other changes:

* <a href="diffhunk://#diff-81c362d167240c77cd39a0678f9bf4de3b93ecd0a7ed07c5eb780173530532ecL41-R41">`integration-test/README.md`</a>: Updated to use the `run.sh` script instead of running the `bats` command directly.
* <a href="diffhunk://#diff-ace417b47e816a44cf3b6f6248e72453a46d9e6043f19aea9d39212e852cc373L5-R11">`main/cmds.go`</a>: Added three import statements for later use.
* Re-target Pull Request to develop/v2 to follow GitFlow.

Hafthor commented on integration-test/test/6_tls-config.bats:
```diff
@@ -0,0 +1,206 @@
+    }' ''
+    run start_container
+    echo "$output"
+    [[ "$output" == *'Grace period set to 10s'* ]]
+    [[ "$output" == *'remote error: tls: protocol version not supported'* ]]
+    [[ "$output" == *'No longer honoring grace period - expired'* ]]
```
what do these lines do? maybe add a comment.

klugorosado replied:
ack

klugorosado replied:
This test verifies the situation where only SSLv3, or outdated and insecure security protocol, is required by the Customers App. New comments have been added.

klugorosado commented on main/health.go:
```diff
@@ -150,10 +154,10 @@ func NewHttpHealthProbe(protocol string, requestPath string, port int) *HttpHeal
-	// remove first slash since we want requestPath to be defined without having to prefix with a slash
+
 	requestPath = strings.TrimPrefix(requestPath, "/")
+	p.Address = fmt.Sprintf("%s://localhost%s/%s", protocol, portString, requestPath)
```
ack

klugorosado commented on integration-test/test/test_helper.bash:
```diff
@@ -6,21 +6,55 @@ TEST_CONTAINER=test
-    echo "Building test image..."
+    if [ -n "$(docker images -q $IMAGE)" ]; then
+        echo "Docker image $IMAGE already exists.">&2
+        local random_number=$((1 + RANDOM % 1000000))
```
ack

frank-pang-msft commented on .github/workflows/go.yml:
```diff
@@ -85,6 +89,20 @@ jobs:
+        sudo bats integration-test/test/1_basic.bats -T --trace
+      working-directory: ${{ env.repo_root }}
+
+    - name: Run Integration Tests in Parallel
```
"Run Parallel Integration Tests"

klugorosado replied:
ack

frank-pang-msft commented on .github/workflows/go.yml:
```diff
@@ -85,6 +89,20 @@ jobs:
 
-    - name: Run Integration Tests
-      run: sudo bats integration-test/test
+    - name: Run Integration Tests Sequentially
```
Run Sequential Integration Tests

klugorosado replied:
ack

frank-pang-msft commented on .github/workflows/go.yml:
```diff
@@ -85,6 +89,20 @@ jobs:
+        sudo bats $(find integration-test/test/ -type f -name "*.bats" ! -name "1_basic.bats") --jobs 10 -T --trace
+      working-directory: ${{ env.repo_root }}
+
+    - name: Retry Failing Integration Tests
```
If Sequential and Parallel tests failed, is that okay? Maybe there should be separate steps for retry sequential and retry parallel?

sequential
retry sequential
parallel
retry parallel

klugorosado replied:
Since this step is to re-test Flaky tests it not necessary to run them in parallel. Normally there is one Flaky test per run (the last to complete) not sure why.

klugorosado commented on integration-test/test/test_helper.bash:
```diff
@@ -6,21 +6,55 @@ TEST_CONTAINER=test
+# a unique name is needed to avoid conflicts with other tests while running in parallel.
 build_docker_image() {
-    echo "Building test image..."
+    if [ -n "$(docker images -q $IMAGE)" ]; then
```
ack

klugorosado commented on integration-test/run.sh:
```diff
@@ -1,3 +1,7 @@
+source integration-test/test/test_helper.bash
+create_certificate
+# 1_basic.bats Integration Test must be run sequentially
+sudo bats integration-test/test/1_basic.bats -T --trace
```
Good recommendation

Hafthor commented on integration-test/test/6_tls-config.bats:
```diff
@@ -0,0 +1,212 @@
+    # ApplicationHealth Extension should only Support from TLS 1.0 to TLS 1.3
+    # if only invalid Security Protocol is supported by the application like SSLv3 
+    # then the extension will receive a Standard TLS error. 
+    [[ "$output" == *'Grace period set to 10s'* ]]
+    [[ "$output" == *'remote error: tls: protocol version not supported'* ]]
+    [[ "$output" == *'No longer honoring grace period - expired'* ]]
+    [[ "$output" == *'Grace period set to 10s'* ]]
+    [[ "$output" == *'remote error: tls: protocol version not supported'* ]]
+    [[ "$output" == *'No longer honoring grace period - expired'* ]]
```
whoops. looks like we have duplicate lines here

klugorosado replied:
No sure how this happened.

Hafthor commented on main/health.go:
```diff
@@ -119,6 +120,27 @@ func (p *TcpHealthProbe) healthStatusAfterGracePeriodExpires() HealthStatus {
+		portString = ":" + strconv.Itoa(port)
+	}
+
+	requestPath = strings.TrimPrefix(requestPath, "/")
```
nit: url.URL already trims leading / from Path

klugorosado replied:
Left it because the documentation says "may omit leading slash" but will change.

Hafthor commented on main/health_test.go:
```diff
@@ -0,0 +1,53 @@
+
+func TestNewHttpHealthProbeWithDefaultHttpPort(t *testing.T) {
+	protocol := "http"
+	requestPath := "/test"
```
maybe make a couple of these tests with "test" as the requestPath to test "/" trimming

klugorosado replied:
ack

Hafthor commented on integration-test/webserver/main.go:
```diff
@@ -147,16 +148,53 @@ func healthHandler(w http.ResponseWriter, r *http.Request, arguments *[]string)
+
+func getHealthState(flag string) string {
+
+	if _, ok := healthStateFlagMapping[flag]; ok {
+		return healthStateFlagMapping[flag]
+	} else {
+		return ""
+	}
```
nit: return the captured healthState

    if healthState, ok := healthStateFlagMapping[flag]; ok {
        return healthState
    } else {
        return ""
    }

frank-pang-msft commented on .github/workflows/go.yml:
```diff
@@ -85,6 +89,20 @@ jobs:
+        sudo bats $(find integration-test/test/ -type f -name "*.bats" ! -name "1_basic.bats") --jobs 10 -T --trace
+      working-directory: ${{ env.repo_root }}
+
+    - name: Retry Failing Integration Tests
```
I guess it's fine, but I noticed in the github actions, sequential and parallel tests do not fail, but the retry step still executes?

klugorosado replied:
just fixed the problem

frank-pang-msft commented:
Can you update PR description with the latest?
